### ISSUE TITLE
Formatting rules for subqueries

### ISSUE THREAD

**Issue by fdw** [sqlfluff/sqlfluff#1862]
Similar to #1861, I would expect sqlfluff to have formatting rules regarding subqueries, for example
```sql
if exists(select 1 from...)
begin
  drop ...
end
```

I personally would prefer the subquery to be in a new, indented line, but it's also okay if it's consistently in the same line. Currently, however, there is no alignment whatsoever.

**Comment by barrywhart** [comment]
Can you provide a full example query? Also, what dialect are you using above?

**Comment by fdw** [comment]
Sure! Here's a fuller example:
```sql
if exists (select * from sys.objects where object_id = object_id(N'[SomeSchema].[ProcedureName]') and type in (N'P'))
begin
  drop procedure SomeSchema.ProcedureName
end
```

The SQL dialect is T-SQL.

**Comment by jpy-git** [comment]
I'd be happy to implement a rule for subqueries that formats them on a new indented line. This approach would work nicely with rule L003 as it would ensure that the subquery segments have correct indentation relative to the outer statement.

So
```SQL
SELECT *
FROM (SELECT * FROM foo)
```
becomes
```SQL
SELECT *
FROM (
    SELECT * FROM foo
)
```

**Comment by barrywhart** [comment]
It sounds good to me. I could see this as a new rule out a new, optional behavior for L003. (Either seems okay from a user perspective. I'm not sure which would be easier to implement.)

**Comment by jpy-git** [comment]
I think a new rule would be simplest, let L003 handle indentation only rather than also moving segments vertically. I'll pick this up tomorrow 🚀

**Comment by barrywhart** [comment]
>I think a new rule would be simplest
That's my guess, too. If necessary, you can copy or otherwise reuse some code from L003. We have a few other rules which do that.

**Comment by fdw** [comment]
What is the plan for long subqueries? Can they be broken down according to L003, but with a larger indent?

**Comment by jpy-git** [comment]
There's already a rule for line too long so that should handle that

**Comment by barrywhart** [comment]
We'll want some tests for that scenario. Idea: Copy several of the existing L003 tests and make the existing query a subquery.

**Comment by jpy-git** [comment]
Update:
I've got this rule working fine for the subquery contents but it seems that L003 (or some other rule) is also trying to mess with subquery bracket position which I don't think we want. L003 is a bit of a beast so I'll need to find some time to figure out what's going on there.

**Comment by barrywhart** [comment]
Good luck! I've had some struggles with understanding L003 as well -- it's probably the most complex of the linting rules (at least that I've touched). As you're looking at it, if you have any thoughts on how we could make it easier to work on, that would be 💯.

A few ideas I've had: add docs, examples, or additional unit tests for specific fields or variables such as:
* `hanging_indent`
* `clean_indent`
* `hanger_pos`
* Others?

As it's currently written, the rule does many things, and it can be difficult to grasp it because one function often contributes to several different behaviors.

Related: I think in the last maintainers meeting, there may have been some discussion about revisiting how formatting works and interacts with the other rules? @alanmcruickshank

**Comment by XiChu2333** [comment]
I encounter some cases similar to this. I think it's more about new line. 
I would expect the parentheses that contain a subquery inside go a new line.

One of my examples look like following: 
```
select A.Something
from (select
        B,
        C
    from Table group by 1) as A
```
It would looks much better like this following
```
select A.Something
from (
    select
        B,
        C
    from Table group by 1
) as A
```

**Comment by XiChu2333** [comment]
The new-line issue in the initial case seems to solved in the new version 1.0. Here is my test:
test.sql
```
if exists (select * from sys.objects where object_id = object_id(N'[SomeSchema].[ProcedureName]') and type in (N'P'))
begin
  drop procedure SomeSchema.ProcedureName
end
```
sqlfluff fix test.sql --dialect tsql
```
if
    exists (
        select *
        from
            sys.objects
        where
            object_id = object_id(
                N'[SomeSchema].[ProcedureName]'
            ) and type in (N'P')
    )
    begin
        drop procedure someschema.procedurename
    end
```
The subquery is been put in a new, indent indented line. But the indentation of the rest is not correct, maybe this is due to tsql dialect?
